### DISCUSSION TITLE
Different design decisions and possible collaboration

### DISCUSSION THREAD

**Discussion by stephanemagnenat** [emilk/egui#847] [Ideas]
Hello,

At ETH Zürich, I am using Egui along with Bevy in a student project and we are very happy about it. In parallel, in a [spin-off I co-founded](https://enlightware.ch), we are developing [a platform](https://cand.li) for people to create their own video game, whose client code is currently written in Typescript with a bit of Rust, with a GUI written in a combination of Google MDC and custom Typescript-based Imgui.

However, our long-term plan is to move to Rust, including the GUI code. We have several features that we are interested in having, and that Egui does not provide. So I have been creating a proof-of-concept GUI library + WebGL2 renderer to explore these features, which are:
- High performance modern renderer using signed-distance function with perfect anti-aliasing and fast scalable font rendering ([demo](https://cand.li/test-unified-renderer.html)), currently lacking texture support besides font rendering.
- Presentation-agnostic widget logic, with a Google Material presentation as an example ([demo](https://cand.li/test-exkui.html)).
- Generate IDs through `#[track_caller]` annotation. This is clearly limiting but did not limit me so far, and one could add an optional explicit ID when needed. 

Beside the very rough state of these proof-of-concepts, there are still features missing we need for our product such as proper recursive drag & drop support as we have in our Typescript-based Imgui library (see the [Candli program editor](https://cand.li)) for an example. However, these early demos show that the transition is doable.

Yet, my goal is not to re-invent the wheel for the sake of reinvention, and if there could be ways to collaborate with Egui I would be very happy. I have been thinking a bit about it, and the problem for us to actually use Egui is that its current design places abstractions at the wrong place for us. For example, Egui currently abstracts the backend at the level of the triangles and textures, which prevents the implementation of something like Google Material style as well as using a more optimised renderer such as the one above.

Nevertheless, parts could still be shared, for example our renderer could be made into a library that could be of use for Egui – or other projects – one day. Our code is currently not open source, mainly because it is really about exploration and I have no intention to support it in its current form, but if there is interest for collaboration, open-sourcing is totally an option middle term.

So I would be very happy to hear what the Egui community thinks about the ideas described above and possible way of collaborating.

**Comment by emilk** [comment]
Cool demo!

> Egui currently abstracts the backend at the level of the triangles and textures

egui actually first outputs a list of [`Shape`](https://docs.rs/epaint/0.15.0/epaint/enum.Shape.html)s: <https://docs.rs/egui/0.15.0/egui/struct.Context.html#method.end_frame>. Tessellation is a separate step!

As for text rendering: I would like to explore some alternatives here, including SDF and vector rendering, but ultimately I don't have the time for that right now. A higher priority is to switch to a 3rd party crate to do the text layout and rendering (e.g. something like [swash](https://github.com/dfrg/swash)) so egui can support all the complexity of unicode without me having to implement it all. Whatever crate we settle on there will likely influence how egui will do text rendering.

**Reply by stephanemagnenat** [reply]
> egui actually first outputs a list of Shapes: https://docs.rs/egui/0.15.0/egui/struct.Context.html#method.end_frame. Tessellation is a separate step!

Right, sorry, I forgot about that indeed. Maybe the most relevant point in that context is that egui combines widget logic and presentation.

Swash looks great indeed. I also have no interest in re-inventing the wheel there. On the short-term, we probably do not need all the unicode glory but on the long run probably we do.

Regardless of text rendering, would you be interested in a shape-rendering back-end that can efficiently and with high quality render basic geometric shapes using sign-distance functions computations done in a pixel shader, as shown in the demo above? That would target GPU feature sets roughly equivalent to what WebGL2 provides. If so, I could imagine repackaging my experiment into something more useful by others.
